File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        queue = collections.deque([])

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 0:
                    queue.append((i,j,0))

        directions = [(1,0),(0,1),(-1,0),(0,-1)]

        def is_valid(m,n):
            return 0 <= m < len(grid) and 0 <= n < len(grid[0])

        while queue:
            level = len(queue)
            for i in range(level):
                cell = queue.popleft()
                for direction in directions:
                    m = cell[0] + direction[0]
                    n = cell[1] + direction[1]
                    if is_valid(m,n) and grid[m][n] == 2147483647:
                        grid[m][n] = 1 + cell[2]
                        queue.append((m,n,grid[m][n]))
            

File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) {
        int islands = 0;
        for(int i = 0; i < grid.length; i ++) {
            for (int j = 0; j < grid[0].length; j++) {
                if(grid[i][j] == '1') {
                    islands++;
                    dfs(grid, i, j);
                }
            }
        }
        return islands;
    }

    public void dfs(char[][] grid, int i , int j) {
        grid[i][j] = '0';
        int m = grid.length;
        int n = grid[0].length;
        if(i + 1 < m && grid[i + 1][j] == '1'){
            dfs(grid, i+1, j);
        }
        if(i - 1 >= 0 && grid[i - 1][j] == '1'){
            dfs(grid, i - 1, j);
        }
        if(j + 1 < n && grid[i][j + 1] == '1'){
            dfs(grid, i, j + 1);
        }
        if(j - 1 >= 0 && grid[i][j - 1] == '1'){
            dfs(grid, i, j - 1);
        }
    }
}
